sw/cheri/checks/pinmux_checker.hh

```diff
@@ -0,0 +1,163 @@
+  uint8_t bit;            // 0-31
+};
+
+inline void set_gpio_output(Capability<volatile SonataGpioFull> gpio, GpioPin pin, bool value);
+inline void set_gpio_output_enable(Capability<volatile SonataGpioFull> gpio, GpioPin pin, bool value);
+inline bool get_gpio_input(Capability<volatile SonataGpioFull> gpio, GpioPin pin);
+
+// Enum representing the different type of tests currently able to be implemented
+// in a pinmux testplan.
+enum class TestType {
+  UartSendReceive     = 0,
+  GpioWriteRead       = 1,
+  I2cBH1745ReadId     = 2,
+  I2cPmodColourReadId = 3,
+  SpiPmodSF3ReadId    = 4,
+};
+
+// The test-specific data required to carry a UART send/receive test
+struct UartTest {
+  enum class UartNum : uint8_t {
+    // Uart0 is used as a console and cannot be tested
+    Uart1 = 1,
+    Uart2 = 2,
+    Uart3 = 3,
+    Uart4 = 4,
+  } uart;
+  uint32_t timeout;
+  uint32_t test_length;
+};
+
+// The test-specific data required to carry out a GPIO write/read test
+struct GpioTest {
+  GpioPin output_pin;
+  GpioPin input_pin;
+  uint32_t wait_time;
+  uint32_t test_length;
+};
+
+// The test-specific data required to carry out an I2C BH1745 Read ID test
+struct I2cTest {
+  enum class I2cNum : uint8_t {
+    I2c0 = 0,
+    I2c1 = 1,
+  } i2c;
+};
+
+// The test-specific data required to carry out a SPI Pmod SF3 Read ID test
+struct SpiTest {
+  enum class SpiNum : uint8_t {
+    // Spi0-2 are the Flash, LCD and Ethernet respectively and cannot be tested.
+    Spi3 = 3,
+    Spi4 = 4,
+  } spi;
+  GpioPin cs_pin;
+};
+
+// An output pin assignment to be made via Pinmux for a test
+struct OutputPinAssignment {
+  SonataPinmux::OutputPin pin;
+  uint8_t select;
+};
+
+// A block input assignment to be made via Pinmux for a test
+struct BlockInputAssignment {
+  SonataPinmux::BlockInput input;
+  uint8_t select;
+};
+
+// A single pinmux test to be executed
+struct Test {
+  // The type of test being specified.
+  TestType type;
+  // The name of the test to print to the console
+  const char *name;
+  // Whether testplan execution should pause for manual intervention. If this
+  // is set to true, then an `instruction` must be specified, which is the
+  // message to print to the console to direct the user.
+  bool manual_required;
+  const char *instruction;
+  // The output pin and block input assignments to pinmux before testing
+  OutputPinAssignment *output_pins;
+  uint32_t num_output_pins;
+  BlockInputAssignment *block_inputs;
+  uint32_t num_block_inputs;
+  /**
+   * TODO: Ideally we would union the below attributes to save space, but the lack of
+   * `memcpy` makes that very difficult, i.e.
+   *   union TestData {
+   *     UartTest uart;
+   *     GpioTest gpio;
+   *     I2cTest i2c;
+   *     SpiTest spi;
+   *   } test_data;
+   */
+  // The data to use in the test. The attribute corresponding to the specified `type`
+  // attribute must be filled in with appropriate data.
```

Comment:
```suggestion
  /**
   * The data to use in the test. The attribute corresponding to the specified `type`
   * attribute must be filled in with appropriate data.
   * Ideally we would union the below attributes to save space, but the lack of
   * `memcpy` makes that very difficult, i.e.
   *   union TestData {
   *     UartTest uart;
   *     GpioTest gpio;
   *     I2cTest i2c;
   *     SpiTest spi;
   *   } test_data;
   */
```

Reply:
Side comment: Ideally I would figure out a more object-oriented approach using C++'s classes here but I spent quite a while battling the lack of `memcpy` and so if this is a change we think is important, I can look at iterating on this later down the line.